floss/identify.py

```diff
@@ -0,0 +1,130 @@
+# Copyright (C) 2017 FireEye, Inc. All Rights Reserved.
+
+import logging
+import operator
+import collections
+
+import tqdm
+import viv_utils
+
+from floss.features.extract import (
+    abstract_features,
+    extract_insn_features,
+    extract_function_features,
+    extract_basic_block_features,
+)
+from floss.features.features import Arguments, BlockCount, InstructionCount
+
+logger = logger = logging.getLogger(__name__)
+
+
+def get_function_api(f):
+    ret_type, ret_name, call_conv, func_name, args = f.vw.getFunctionApi(int(f))
+
+    return {
+        "ret_type": ret_type,
+        "ret_name": ret_name,
+        "call_conv": call_conv,
+        "func_name": func_name,
+        "arguments": args,
+    }
+
+
+def get_function_meta(f):
+    meta = f.vw.getFunctionMetaDict(int(f))
+
+    return {
+        "api": get_function_api(f),
+        "size": meta.get("Size", 0),
+        "block_count": meta.get("BlockCount", 0),
+        "instruction_count": meta.get("InstructionCount", 0),
+    }
+
+
+def get_max_calls_to(vw, skip_thunks=True, skip_libs=True):
+    calls_to = set()
+
+    for fva in vw.getFunctions():
+        if skip_thunks and is_thunk_function(vw, fva):
+            continue
+
+        # TODO skip_libs and is_library_function
+        #     continue
+
+        calls_to.add(len(vw.getXrefsTo(fva)))
+
+    return max(calls_to)
+
+
+def is_thunk_function(vw, function_address):
+    return vw.getFunctionMetaDict(function_address).get("Thunk", False)
+
+
+def get_function_score_weighted(features):
+    return sum(feature.weighted_score() for feature in features) / sum(feature.weight for feature in features)
```

Comment:
this provides similar results to what we did before, with some data analysis we could figure our feature weights, but I think that's overkill

Reply:
i wonder if we could do a large scale analysis to learn the weights/scores that we should use.

in theory, the filtering is only for performance (to cut down on the number of things we emulate). we could do a heavy, one-time run across many malware samples and identify decoding routines by whether or not we find any results. then, work backwards to find the weights. #ai #ml